Add a new stretch mode for 2-site Ceph clusters

gregsfortytwo opened the pull request:
This PR adds a new "stretch mode" to RADOS.

It targets 2-site clusters (with a tiebreaker mon elsewhere) and extends the existing rules
so that we can directly guarantee peers in each site, and handles netsplits between them.
More details are in the included dev and user documentation.

gregsfortytwo commented:
This PR obviously doesn't meet our usual testing standards. :( But I'd like to get it in-tree early for downstream development reasons — this is directly driven by Red Hat's product goals. I've used the set_up_stretch_mode.sh script and manual testing to verify that we do correctly transition through the various healthy/degraded/recovery stretch modes, and that the appropriate peering rules are enforced at a basic level. I am reasonably confident that this won't break any existing functionality, but waiting on the lab to catch up to run a suite through for that.

gregsfortytwo commented:
This PR is built on a (trivially-rebased) https://github.com/ceph/ceph/pull/32336; we can close that one and merge this if we're happy doing so.

athanatos commented on src/osd/OSDMap.cc:
```diff
@@ -693,7 +694,19 @@ void OSDMap::Incremental::encode(ceph::buffer::list& bl, uint64_t features) cons
     }
-    ENCODE_FINISH(bl); // osd-only data
+    if (target_v >= 10) {
+      encode(change_stretch_mode, bl);
```
Looks to me like the other fields are only interpretted if change_stretch_mode, could we choose not to encode/decode them if !change_stretch_mode?

gregsfortytwo replied:
Talked this over and sounds like we don't want to — Sam has visions of switching normal peering to use some of these fields to solve our mis-placement troubles for Pacific.

athanatos replied:
Indeed, at least peering_crush_bucket_barrier (which should be renamed peering_crush_failure_domain) should be totally independent if we wish to make use of it normally.

athanatos commented on src/osd/PeeringState.cc:
```diff
@@ -5369,6 +5516,7 @@ PeeringState::Recovering::react(const RequestBackfill &evt)
   if (!ps->async_recovery_targets.empty()) {
     pg_shard_t auth_log_shard;
     bool history_les_bound = false;
+    // FIXME: Uh-oh we have to check this return value; choose_acting can fail!
```
Can it?  Our current acting set (and up set, presumably?) meets the requirements, so moving async recovery/backfill targets into acting should always work, right?

gregsfortytwo replied:
You’re probably right; I didn’t evaluate these too carefully, it was more of a note to check. Will look more carefully at what’s going on (though if it’s not possible to fail shouldn’t we assert that?)

athanatos commented on src/osd/osd_types.h:
```diff
@@ -1380,7 +1380,16 @@ struct pg_pool_t {
+  // The per-bucket replica count is calculated with this "target"
+  // instead of the above crush_bucket_count. This means we can maintain a
+  // target size of 4 without attempting to place them all in 1 DC
+  uint32_t peering_crush_bucket_target = 0;
```
s/peering_crush_bucket_count/peering_crush_bucket_min_size
s/peering_crush_bucket_target/peering_crush_bucket_target_size

athanatos commented on src/osd/osd_types.h:
```diff
@@ -1380,7 +1380,16 @@ struct pg_pool_t {
+  // of this bucket type...
+  uint32_t peering_crush_bucket_barrier = 0;
+  // including this one
+  int32_t peering_crush_mandatory_member = 0;
```
I may have missed something, but for this to be effective, PastIntervals::check_new_interval needs to consider this when dealing with maybe_went_rw.

athanatos replied:
Same with bucket_count/barrier.

gregsfortytwo commented:
Will have patches for check_new_interval and the encoding make check failures later today.
Sounds like @athanatos has a new better implementation of calc_replicated_acting incoming as well.

athanatos commented on src/osd/osd_types.h:
```diff
@@ -1380,7 +1380,16 @@ struct pg_pool_t {
+  // of this bucket type...
+  uint32_t peering_crush_bucket_barrier = 0;
+  // including this one
+  int32_t peering_crush_mandatory_member = 0;
```
I don't think you want to use 0 as default/null here.  osd.0 is actually valid.  CRUSH_ITEM_NONE seems like it would be the right answer.

athanatos commented:
I've pushed possibly a more robust approach to calc_replicated_acting for stretch clusters: https://github.com/athanatos/ceph/tree/sjust/wip-stretch-peering .  Note that this is entirely 100% untested -- it needs unit testing and debugging.

I think the above approach generalizes nicely to fixing the existing deficiency with choose_acting in that it doesn't necessarily respect failure domains when constructing a temp mapping.  If we allow peering_crush_bucket_barrier to be used independently of bucket_count and rename it to something like peering_crush_failure_domain, we could set it on all clusters and eliminate the previous implementation entirely in the future.

jdurgin commented on src/mon/MonCommands.h:
```diff
@@ -468,6 +469,32 @@ COMMAND("mon set-weight "
+	"name=args,type=CephString,n=N,goodchars=[A-Za-z0-9-_.=]",
+	"specify location <args> for the monitor <name>, using CRUSH bucket names", \
+	"mon", "rw")
+COMMAND("mon enable_stretch_mode " \
```
is it possible to disable stretch mode?

jdurgin commented on src/common/options.cc:
```diff
@@ -1636,6 +1636,18 @@ std::vector<Option> get_global_options() {
+    .set_default(2)
+    .add_service("mon")
+    .set_description("")
+    .set_min_max(2, 4),
```
could we avoid hardcoding these limits? In some disaster recovery cases we may need to reduce the size/min_size of the pools temporarily, and it'd be nice not to require a new build to do so.

gregsfortytwo replied:
These are just the defaults that get set when you go in to stretch mode; you can still set a specific pool however you want later when doing recovery.

That said the recovery out of stretch mode is going to be complicated by the crush bucket distribution requirements, which I imagine will be more of a barrier than the min size.

jdurgin commented on src/osd/OSDMap.h:
```diff
@@ -552,8 +552,8 @@ class OSDMap {
-    CEPH_FEATUREMASK_SERVER_OCTOPUS;
-  // TODO: add a SIGNIFICANT_FEATURE for stretch mode...
+    CEPH_FEATUREMASK_SERVER_OCTOPUS |
+    CEPH_FEATUREMASK_STRETCH_MODE;
```
I think we need a osdmap flag as well to determine when it's safe to enable stretch mode, so that we can prevent osds without that flag from booting and using an incorrect choose_acting() algorithm. see pg_log_hardlimit for a similar situation - just described for another issue here: https://github.com/ceph/ceph/pull/35938#issuecomment-656183558

gregsfortytwo replied:
Hmm I thought adding this to significant features would do that (it will certainly prevent OSDs from decoding the various structs correctly, as when crush peering values are set the encoding compatv == struct_v!).

It looks like the pg_log_hardlimit does something in the OSDMonitor to prevent OSDs booting in at all, but don't we have OSDMap flags somewhere to handle this case based on the required feature set? It looks like maybe adding it to OSDMap::get_features() for the OSD entity type?

athanatos commented:
@jdurgin Do you mean extracting the logic for primary selection?  Other than that, the existing calc_acting logic is left in its own method for non-stretch clusters.

jdurgin commented:
@athanatos ah that is more minimal than I thought, I was confused by the changes being on top of greg's choose_acting changes. Looking at the diff from master it's clear the stretch_mode pieces are all separate, so it looks good to me.

neha-ojha commented:
@gregsfortytwo https://pulpito.ceph.com/nojha-2020-07-11_00:28:22-rados-wip-35906-distro-basic-smithi/ needs some work

gregsfortytwo commented on src/mon/MonCommands.h:
```diff
@@ -488,10 +488,9 @@ COMMAND("mon set_location " \
+	"name=dividing_bucket,type=CephString, ",
 	"enable stretch mode, changing the peering rules and "
-	"failure handling on the named <pools> with <tiebreaker_mon> "
+	"failure handling on all pools with <tiebreaker_mon> "
```
The data structures support setting this on individual pools and my initial draft had users specify the stretch pools when enabling stretch mode, but it opens up a whole category of new failure cases and UX issues by doing so that I thought we should leave out initially. For instance, when stretch mode is enabled on all pools and somebody creates a new pool, do we assume it's a stretch pool, or make them specify it? Are we comfortable with a situation where they might accidentally have their rgw metadata on a non-stretch pool that gets lost due to a DC failure?

While I can see a future where we want to do it on a per-pool basis, doing it on the whole cluster fits the initial use cases and is a lot simpler/safer.

gregsfortytwo commented:
Pretty big updates here; built on @athanatos' rework of the calc_replicated_acting_stretch, implemented the interval change logic (whoops!), and switched to use CRUSH_ITEM_NONE for the mandatory member as requested. Fixed the MonCommands syntax error which caused all the failures I found in spot checks from the run @neha-ojha did; ran it through last night and popped up an issue with the last bugfix in the older ElectionLogic changes so fixed that as well.

I'm now looking at the OSDMap encoding feature bits @jdurgin mentioned because there was an issue there in an upgrade test (during upgrade from Nautilus; looks like popping it into the SIGNIFICANT_FEATURES and encoding for Nautilus busted it even though none of the features were in use, so definitely need to update that), and adding a command to disable stretch mode. But mostly waiting for Shaman to build so I can run the fixes through qa suites and see if there are any unidentified problems...

gregsfortytwo commented:
https://pulpito.ceph.com/gregf-2020-07-17_02:45:59-rados-wip-stretch-mode-3-distro-basic-smithi/

Almost all the failures are with cephadm and look to me like the container it wants simply isn’t present; is there a separate step that needs to happen to make those work? Something we can track in shaman to see when it’s done?

Haven’t been through the rest carefully yet but I think it’s the osdmap encoding for nautilus daemons I mentioned previously.

gregsfortytwo commented:
Updated the OSDMap feature bit handling; removed it from SIGNIFICANT_FEATURES as that was not actually what we want. The OSDMonitor now additionally prevents boot of OSDs without CEPH_FEATUREMASK_STRETCH_MODE when it's enabled in the OSDMap, and OSDMap::get_features() returns it as a needed OSD feature.

Poked a little more at the ElectionLogic first-election zero-score handling so it now passes make check (whoops about that last push, though I don't think it was likely to show up anywhere outside the unit test framework), and I think I fixed the syntax errors in the docs. More tests should happen overnight, though I had to filter out cephadm as they currently seem to be broken in master.

gregsfortytwo commented:
https://pulpito.ceph.com/gregf-2020-07-20_09:30:49-rados-wip-stretch-mode-4-distro-basic-smithi/

Two jobs hung so I killed them. 5243222 was running ceph_test_msgr and the process was stuck in the MessengerTest.ConnectionRaceReuseBannerTest; 5243339 seemed to have a network hiccup and then the DaemonWatchdog tried and failed to kill the test (I see a hangup coming from osd.0 despite it running happily when I looked, and an issue getting command descriptions from osd.6 but no other indications it had failed).

Of the failures:
5243114: Nautilus OSDs are crashing in a segfault somewhere in/under OSD::_committed_osd_maps. I see they are also failing to encode OSD maps with the expected CRC after getting an incremental, but not sure if this is related.
5243163: a mon election interval where mon.c got left out?
5243172: cephadm failures
5243221: OSD incorrectly marked down, but no Tracebacks.
5243224: a mon election interval where one monitor (of 21) got left out, but no Tracebacks.
5243241: cephadm failure
5243272: cephadm failure
5243287: OSDMap commit crashes
5243406: cephadm failure
5243441: cephadm failure

I think the upgrade failures a result of more encoding issues — current code is unconditionally encoding the new stretch values in pg_pool_t and the OSDMap, assuming the features support it. But that doesn't work when the cluster includes a mix of versions — whoops!

gregsfortytwo commented:
Hmm yep my local testing of a patch to fix the encoding is turning up the same bug now. Besides the immediate mismatche-CRC bug, at first glance this looks to me like the OSD's handling of bad CRCs and fetching full maps is broken. :(

jdurgin commented:
latest updates look good to me

jdurgin commented:
@gregsfortytwo how are the tests looking now, is this ready to go?

gregsfortytwo commented:
@jdurgin @neha-ojha I still have more testing I want to write for this but it's looking good against our current suite:
https://pulpito.ceph.com/gregf-2020-09-14_05:25:36-rados-wip-stretch-mode-distro-basic-smithi/
and I think it needs to go in before more conflicts get introduced (and, well, so I can put it into our downstream builds for testing).

I tracked all the failures/hangs down to identified issues in the tracker, except for creating https://tracker.ceph.com/issues/47453 -- and one where the initial monitor quorum startup took a couple rounds to quiesce, so it flagged MON_DOWN in the log.

neha-ojha commented:
> @jdurgin @neha-ojha I still have more testing I want to write for this but it's looking good against our current suite:
> https://pulpito.ceph.com/gregf-2020-09-14_05:25:36-rados-wip-stretch-mode-distro-basic-smithi/
> and I think it needs to go in before more conflicts get introduced (and, well, so I can put it into our downstream builds for testing).
> 
> I tracked all the failures/hangs down to identified issues in the tracker, except for creating https://tracker.ceph.com/issues/47453 -- and one where the initial monitor quorum startup took a couple rounds to quiesce, so it flagged MON_DOWN in the log.

@gregsfortytwo I am not seeing any new failures in the FAILED jobs, although some have already been fixed in latest master. Among the DEAD jobs, https://tracker.ceph.com/issues/47453 is definitely new, need to check with @ifed01 and 5433261/5433382 are also failing abruptly but don't think this PR is related.

gregsfortytwo commented:
> @gregsfortytwo I am not seeing any new failures in the FAILED jobs, although some have already been fixed in latest master. Among the DEAD jobs, https://tracker.ceph.com/issues/47453 is definitely new, need to check with @ifed01 and 5433261/5433382 are also failing abruptly but don't think this PR is related.

@neha-ojha, is that an approval? I need you or @jdurgin to actually click the review button! :)

ifed01 commented:
> > @jdurgin @neha-ojha I still have more testing I want to write for this but it's looking good against our current suite:
> > https://pulpito.ceph.com/gregf-2020-09-14_05:25:36-rados-wip-stretch-mode-distro-basic-smithi/
> > and I think it needs to go in before more conflicts get introduced (and, well, so I can put it into our downstream builds for testing).
> > I tracked all the failures/hangs down to identified issues in the tracker, except for creating https://tracker.ceph.com/issues/47453 -- and one where the initial monitor quorum startup took a couple rounds to quiesce, so it flagged MON_DOWN in the log.
> 
> @gregsfortytwo I am not seeing any new failures in the FAILED jobs, although some have already been fixed in latest master. Among the DEAD jobs, https://tracker.ceph.com/issues/47453 is definitely new, need to check with @ifed01 and 5433261/5433382 are also failing abruptly but don't think this PR is related.

Honestly I have no clue where https://tracker.ceph.com/issues/47453 comes from. Neither have any idea how to catch it... Would QA re-run reveal it again?
Generally we're getting complains about corrupted RocksDB from time to time for various Ceph releases so it's rather not a recent regression. No evidence whether all of them have the same root cause though...
E.g. here is another log I got yesterday (note that's a monitor's and octopus release, hence no bluefs and lower stuff):
debug 2020-09-15T12:49:48.523+0000 7fdb85d8f700 -1 rocksdb: submit_common error: Corruption: block checksum mismatch: expected 3388771841, got 3472779923  in /var/lib/ceph/mon/ceph-b/store.db/005294.sst offset 108089 size 3925 code = 2 Rocksdb transaction: 
Put( Prefix = m key = 'n_sync'0x006c6174'est_monmap' Value size = 379)
Put( Prefix = m key = 'n_sync'0x00696e5f'sync' Value size = 8)
Put( Prefix = m key = 'n_sync'0x006c6173't_committed_floor' Value size = 8)
/home/abuild/rpmbuild/BUILD/ceph-15.2.4-827-g318de690ed/src/mon/MonitorDBStore.h: In function 'int MonitorDBStore::apply_transaction(MonitorDBStore::TransactionRef)' thread 7fdb85d8f700 time 2020-09-15T12:49:48.529255+0000
/home/abuild/rpmbuild/BUILD/ceph-15.2.4-827-g318de690ed/src/mon/MonitorDBStore.h: 354: ceph_abort_msg("failed to write to db")
 ceph version 15.2.4-827-g318de690ed (318de690ed071d190839b6c05d618915be59b05f) octopus (stable)
 1: (ceph::__ceph_abort(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0xe1) [0x7fdb92e284c9]
 2: (MonitorDBStore::apply_transaction(std::shared_ptr<MonitorDBStore::Transaction>)+0x107f) [0x55af70694b3f]
 3: (Monitor::sync_start(entity_addrvec_t&, bool)+0x1ec) [0x55af706a38ec]
 4: (Monitor::handle_probe_reply(boost::intrusive_ptr<MonOpRequest>)+0xc01) [0x55af706be611]
 5: (Monitor::handle_probe(boost::intrusive_ptr<MonOpRequest>)+0x1af) [0x55af706bfd1f]
 6: (Monitor::dispatch_op(boost::intrusive_ptr<MonOpRequest>)+0x10c8) [0x55af706d4868]
 7: (Monitor::_ms_dispatch(Message*)+0x4fa) [0x55af706d513a]
 8: (Dispatcher::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x58) [0x55af707032b8]
 9: (DispatchQueue::entry()+0x11c2) [0x7fdb93019d42]
 10: (DispatchQueue::DispatchThread::entry()+0xd) [0x7fdb930b87bd]
 11: (()+0x84f9) [0x7fdb91c7c4f9]
 12: (clone()+0x3f) [0x7fdb90e7ffbf]

jdurgin reviewed (approved):
update looks good

neha-ojha commented:
> > @gregsfortytwo I am not seeing any new failures in the FAILED jobs, although some have already been fixed in latest master. Among the DEAD jobs, https://tracker.ceph.com/issues/47453 is definitely new, need to check with @ifed01 and 5433261/5433382 are also failing abruptly but don't think this PR is related.
> 
> @neha-ojha, is that an approval? I need you or @jdurgin to actually click the review button! :)

Actually, I'd like to run this PR on top of latest master, since your upgrade test runs are encountering a bug that's now fixed.

yuriw commented:
![](https://github.trello.services/images/mini-trello-icon.png) [wip-yuri7-testing-2020-09-16-1533-master](https://trello.com/c/KIhJ6KUv/1012-wip-yuri7-testing-2020-09-16-1533-master)

gregsfortytwo commented:
Okay, Yuri's run had 2 upgrade tests fail but they were both due to finding CephFS client evictions in the logs; things still ran to completion.

I pushed a final new patch to deal with the issue in ceph_test_argparse.py that was discovered in backporting to nautilus, and I created a ticket so that we can prevent discovering things at that late stage in future: https://tracker.ceph.com/issues/47509

neha-ojha commented:
> > > @gregsfortytwo I am not seeing any new failures in the FAILED jobs, although some have already been fixed in latest master. Among the DEAD jobs, https://tracker.ceph.com/issues/47453 is definitely new, need to check with @ifed01 and 5433261/5433382 are also failing abruptly but don't think this PR is related.
> > 
> > 
> > @neha-ojha, is that an approval? I need you or @jdurgin to actually click the review button! :)
> 
> Actually, I'd like to run this PR on top of latest master, since your upgrade test runs are encountering a bug that's now fixed.

@gregsfortytwo  https://pulpito.ceph.com/yuriw-2020-09-16_17:34:03-rados-wip-yuri7-testing-2020-09-16-1533-master-distro-basic-smithi/ - the upgrade tests look good and no other related failures, go ahead with the merge.
